openmmml/models/nequippotential.py

```diff
@@ -0,0 +1,410 @@
+copy of this software and associated documentation files (the "Software"),
+to deal in the Software without restriction, including without limitation
+the rights to use, copy, modify, merge, publish, distribute, sublicense,
+and/or sell copies of the Software, and to permit persons to whom the
+Software is furnished to do so, subject to the following conditions:
+
+The above copyright notice and this permission notice shall be included in
+all copies or substantial portions of the Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
+THE AUTHORS, CONTRIBUTORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM,
+DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR
+OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE
+USE OR OTHER DEALINGS IN THE SOFTWARE.
+"""
+
+from typing import Iterable, List, Optional, Tuple
+
+import openmm
+
+from openmmml.mlpotential import MLPotential, MLPotentialImpl, MLPotentialImplFactory
+
+
+class NequIPPotentialImplFactory(MLPotentialImplFactory):
+    """This is the factory that creates NequipPotentialImpl objects."""
+
+    def createImpl(
+        self,
+        name: str,
+        modelPath: str,
+        lengthScale: float,
+        energyScale: float,
+        atomTypes: Optional[List[int]] = None,
+    ) -> MLPotentialImpl:
+        return NequIPPotentialImpl(name, modelPath, lengthScale, energyScale, atomTypes)
+
+
+class NequIPPotentialImpl(MLPotentialImpl):
+    """
+    This is the MLPotentialImpl implementing support for E(3)-equivariant
+    interatomic potentials generated by NequIP or Allegro.
+
+    The potential must be constructed using the NequIP/Allegro code to build and
+    deploy a PyTorch model, which can then be integrated into OpenMM using a
+    TorchForce. The TorchForce is used to add it to the OpenMM System. Note that
+    you must provide a deployed model, as no general purpose model is available.
+
+    To use a deployed model in openmm-ml, you must provide the path to the model,
+    and the conversion factors between the model length and energy units and OpenMM
+    units (nm and kJ/mol, respectively). For example:
+
+    >>> potential = MLPotential('nequip',
+                                modelPath='example_model_deployed.pth',
+                                lengthScale=0.1, # Angstrom to nm
+                                energyScale=4.184 # kcal/mol to kJ/mol
+                                )
+
+    Additionally, you can specify the atom types for the model if the model was
+    trained with custom atom types by passing a list of atom types to the
+    ``atomTypes`` parameter. This must be a list containing an integer specifying
+    the atom type of each particle in the system. Note that by default the model
```

Comment:
Based on the code, I think this description is incorrect.  It actually should contain the atom type for each particle that will be modeled with the ML potential.  So if you call `createMixedSystem()`, it should contain one element for each element of the `atoms` argument, not one for each particle in the System.  Can you make this clear both here and in the description of the `atomTypes` argument below?

Reply:
Thanks for noticing this. I have now clarified it. Now, `atomTypes` is also an argument passed during system creation. This allows creating systems with varying ML regions from the same MLPotential in cases where custom nequip atom types are being used.